### ISSUE TITLE
[BUG] @With generates non-compiling code under Eclipse

### ISSUE THREAD

**Issue by Gaibhne** [rzwitserloot/lombok#2571]
**Describe the bug**
Note that the entire issue is about compilation in Eclipse; I am not able to reproduce the issue via Maven.

@With cannot be used on an inner class where the outer class is generic (I think). An error message such as this is generated when compiled with a recent Java version (used to work on 8):

`Cannot allocate the member type TestCase<T>.Inner using a parameterized compound name; use its simple name and an enclosing instance of type TestCase<T>`

**To Reproduce**
The following code, compiled with Java 11, generates the above error:

```
import lombok.AllArgsConstructor;
import lombok.With;

public class TestCase<T> {
        @AllArgsConstructor
	public class Inner {
		@With
		Object object;
	}
}
```

**Expected behavior**
I expect this to compile under Eclipse, as it used to when running with JDK 8.

**Version info (please complete the following information):**
 - Lombok version: 1.18.12
 - Eclipse: Version: 2020-06 (4.16.0) / Build id: 20200615-1200 (I do not believe this is relevant, as it occurs with multiple Eclipse versions)

**Comment by Rawi01** [comment]
You are right, it does not work. Actually this one is quite interesting, even the delomboked result fails in eclipse while it works in `javac`. I will check out how we can generate code that eclipse likes too.

Delomboked code:
```java
public class WithInnerGeneric<Z> {
	class Inner {
		final String x;

		@java.lang.SuppressWarnings("all")
		public Inner(final String x) {
			this.x = x;
		}

		@java.lang.SuppressWarnings("all")
		public WithInnerGeneric<Z>.Inner withX(final String x) {
			return this.x == x ? this : new WithInnerGeneric<Z>.Inner(x);
		}
	}
}
}
```

**Comment by Rawi01** [comment]
It should be possible to always call the constructor using the nested class name instead of also adding the parent class. This seems to work in all cases I tried. I will change the behaviour and add test cases to cover all the different cases.

**Comment by rzwitserloot** [comment]
We can't generate just `public Inner withX() {}` - see issue #2268 for why that's problematic (and why this until recently did work; the fix for #2268 is recent and added the more partially qualified names to the output of `@With` and `@Builder` and such. We'll need to investigate if ecj is right or javac is right (usually ecj is right, but I do have my doubts here), and if there's anything we can generate that satisfies both ecj and javac and still addresses #2268. Tricky one.

**Comment by Rawi01** [comment]
Uh, I totally missed that bug, sorry about that. I'm courious why all test cases still pass, I guess I have to understand #2268 completly but I'm to tired for that right now. One thing I noticed is that I only changed the constructor call to use the simple name and the returned type is still `Outer.Inner`.

**Comment by rspilker** [comment]
@Rawi01 I don't think we ever created a test for #2268.

**Comment by Rawi01** [comment]
Yes, I added one and think that I found a solution, I need some more time to properly test all the different cases.
